// @carbon/ibm-cloud-cognitive/es/components/PageHeader/PageHeader.js
import _extends from "@babel/runtime/helpers/extends";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["actionBarItems", "availableSpace", "background", "breadcrumbItems", "className", "collapseHeader", "collapseHeaderLabel", "expandHeaderLabel", "collapseHeaderToggleWanted", "preventBreadcrumbScroll", "navigation", "pageActions", "pageHeaderOffset", "preCollapseTitleRow", "subtitle", "tags", "title", "titleIcon"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

//
// Copyright IBM Corp. 2020, 2020
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//
import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import { extractShapesArray } from '../../global/js/utils/props-helper';
import { layout05, baseFontSize } from '@carbon/layout';
import cx from 'classnames';
import ReactResizeDetector from 'react-resize-detector';
import { useWindowResize, useWindowScroll } from '../../global/js/use';
import { BreadcrumbItem, Grid, Column, Row, Button, SkeletonText, Tag } from 'carbon-components-react';
import { ActionBar } from '../ActionBar/';
import { BreadcrumbWithOverflow } from '../BreadcrumbWithOverflow';
import { TagSet } from '../TagSet';
import { ButtonSetWithOverflow } from '../ButtonSetWithOverflow';
import { pkg } from '../../settings';
import { ChevronUp16 } from '@carbon/icons-react';
import { deprecateProp, deprecatePropUsage, prepareProps } from '../../global/js/utils/props-helper';
var componentName = 'PageHeader';
import { blockClass, utilCheckUpdateVerticalSpace, utilGetTitleShape, utilSetCustomCSSProps } from './PageHeaderUtils';
export var PageHeader = /*#__PURE__*/React.forwardRef(function (_ref, ref) {
  var _extractShapesArray, _ref5, _cx, _ref7, _cx4, _cx9;

  var actionBarItems = _ref.actionBarItems,
      availableSpace = _ref.availableSpace,
      background = _ref.background,
      breadcrumbItems = _ref.breadcrumbItems,
      className = _ref.className,
      collapseHeader = _ref.collapseHeader,
      collapseHeaderLabel = _ref.collapseHeaderLabel,
      expandHeaderLabel = _ref.expandHeaderLabel,
      collapseHeaderToggleWanted = _ref.collapseHeaderToggleWanted,
      preventBreadcrumbScroll = _ref.preventBreadcrumbScroll,
      navigation = _ref.navigation,
      pageActions = _ref.pageActions,
      pageHeaderOffset = _ref.pageHeaderOffset,
      preCollapseTitleRow = _ref.preCollapseTitleRow,
      subtitle = _ref.subtitle,
      tags = _ref.tags,
      title = _ref.title,
      titleIcon = _ref.titleIcon,
      rest = _objectWithoutProperties(_ref, _excluded);

  var _useState = useState({}),
      _useState2 = _slicedToArray(_useState, 2),
      metrics = _useState2[0],
      setMetrics = _useState2[1]; // refs


  var localHeaderRef = useRef(null);
  var headerRef = ref || localHeaderRef; // utility functions
  // Title shape is used to allow title to be string or shape

  var getTitleShape = function getTitleShape() {
    return utilGetTitleShape(title, titleIcon, PageHeader.defaultProps.title);
  };

  var checkUpdateVerticalSpace = function checkUpdateVerticalSpace() {
    return utilCheckUpdateVerticalSpace(headerRef, navigation, preventBreadcrumbScroll, setMetrics);
  }; // state based on props only


  var actionBarItemArray = extractShapesArray(actionBarItems);
  var hasActionBar = actionBarItemArray.length;
  var hasBreadcrumbRow = !(breadcrumbItems === undefined && actionBarItems === undefined);
  var pageActionsItemArray = (_extractShapesArray = extractShapesArray(pageActions)) === null || _extractShapesArray === void 0 ? void 0 : _extractShapesArray.map(function (shape) {
    return _objectSpread({
      label: shape.children
    }, shape);
  });
  /* Title shape is used to allow title to be string or shape */

  var titleShape = getTitleShape(); // NOTE: The buffer is used to add space between the bottom of the header and the last content
  // Not pre-collapsed and (subtitle or available space)

  var lastRowBufferActive = (title || pageActions) && !preCollapseTitleRow || subtitle || availableSpace; // state based on scroll/resize based effects

  var _useState3 = useState(false),
      _useState4 = _slicedToArray(_useState3, 2),
      pageActionsInBreadcrumbRow = _useState4[0],
      setPageActionsInBreadcrumbRow = _useState4[1];

  var _useState5 = useState(0),
      _useState6 = _slicedToArray(_useState5, 2),
      scrollYValue = _useState6[0],
      setScrollYValue = _useState6[1];

  var _useState7 = useState(0),
      _useState8 = _slicedToArray(_useState7, 2),
      backgroundOpacity = _useState8[0],
      setBackgroundOpacity = _useState8[1];

  var _useState9 = useState(false),
      _useState10 = _slicedToArray(_useState9, 2),
      hasCollapseButton = _useState10[0],
      setHasCollapseButton = _useState10[1];

  var _useState11 = useState(false),
      _useState12 = _slicedToArray(_useState11, 2),
      spaceForCollapseButton = _useState12[0],
      setSpaceForCollapseButton = _useState12[1];

  var _useState13 = useState(0),
      _useState14 = _slicedToArray(_useState13, 2),
      actionBarMaxWidth = _useState14[0],
      setActionBarMaxWidth = _useState14[1];

  var _useState15 = useState(0),
      _useState16 = _slicedToArray(_useState15, 2),
      actionBarMinWidth = _useState16[0],
      setActionBarMinWidth = _useState16[1];

  var _useState17 = useState(0),
      _useState18 = _slicedToArray(_useState17, 2),
      pageActionInBreadcrumbMaxWidth = _useState18[0],
      setPageActionInBreadcrumbMaxWidth = _useState18[1];

  var _useState19 = useState(0),
      _useState20 = _slicedToArray(_useState19, 2),
      pageActionInBreadcrumbMinWidth = _useState20[0],
      setPageActionInBreadcrumbMinWidth = _useState20[1];

  var _useState21 = useState(0),
      _useState22 = _slicedToArray(_useState21, 2),
      actionBarColumnWidth = _useState22[0],
      setActionBarColumnWidth = _useState22[1];

  var _useState23 = useState(false),
      _useState24 = _slicedToArray(_useState23, 2),
      fullyCollapsed = _useState24[0],
      setFullyCollapsed = _useState24[1]; // handlers


  var handleActionBarWidthChange = function handleActionBarWidthChange(_ref2) {
    var minWidth = _ref2.minWidth,
        maxWidth = _ref2.maxWidth;

    /* don't know how to test resize */

    /* istanbul ignore next */
    setActionBarMaxWidth(maxWidth);
    /* don't know how to test resize */

    /* istanbul ignore next */

    setActionBarMinWidth(minWidth);
  };

  var handleButtonSetWidthChange = function handleButtonSetWidthChange(_ref3) {
    var minWidth = _ref3.minWidth,
        maxWidth = _ref3.maxWidth;

    /* don't know how to test resize */

    /* istanbul ignore next */
    setPageActionInBreadcrumbMaxWidth(maxWidth);
    /* don't know how to test resize */

    /* istanbul ignore next */

    setPageActionInBreadcrumbMinWidth(minWidth);
  };

  var handleResizeActionBarColumn = function handleResizeActionBarColumn(width) {
    /* don't know how to test resize */

    /* istanbul ignore next */
    setActionBarColumnWidth(width);
  };

  var handleResize = function handleResize() {
    // receives width and height parameters if needed

    /* don't know how to test resize */

    /* istanbul ignore next */
    checkUpdateVerticalSpace();
  };

  var toggleCollapse = function toggleCollapse(forceCollapse) {
    var collapse = typeof forceCollapse !== 'undefined' ? forceCollapse : !fullyCollapsed;
    /* don't know how to test resize */

    /* istanbul ignore next if */

    if (collapse) {
      window.scrollTo({
        top: pageHeaderOffset - ((metrics === null || metrics === void 0 ? void 0 : metrics.headerTopValue) || 0),
        behavior: 'smooth'
      });
    } else {
      window.scrollTo({
        top: 0,
        behavior: 'smooth'
      });
    }
  };

  var handleCollapseToggle = function handleCollapseToggle() {
    toggleCollapse();
  }; // use effects


  useEffect(function () {
    // Determine the location of the pageAction buttons
    setPageActionsInBreadcrumbRow(preCollapseTitleRow || scrollYValue > metrics.titleRowSpaceAbove && hasActionBar);
  }, [hasActionBar, metrics.breadcrumbRowSpaceBelow, metrics.titleRowSpaceAbove, preCollapseTitleRow, scrollYValue]);
  useEffect(function () {
    var _utilSetCustomCSSProp;

    var newActionBarWidth = 'initial';
    var newPageActionInBreadcrumbWidth = 'initial';
    /* don't know how to test resize */

    /* istanbul ignore next if */

    if (actionBarColumnWidth > 0) {
      if (pageActionInBreadcrumbMaxWidth > 0 && actionBarColumnWidth > actionBarMaxWidth + pageActionInBreadcrumbMaxWidth) {
        newPageActionInBreadcrumbWidth = "".concat(pageActionInBreadcrumbMaxWidth, "px");
      } else if (pageActionInBreadcrumbMinWidth > 0) {
        newPageActionInBreadcrumbWidth = "".concat(pageActionInBreadcrumbMinWidth, "px");
      }

      if (actionBarMaxWidth > 0 && actionBarColumnWidth > pageActionInBreadcrumbMinWidth + actionBarMaxWidth) {
        newActionBarWidth = "".concat(actionBarMaxWidth, "px");
      } else {
        if (actionBarMinWidth > 0) {
          newActionBarWidth = "".concat(actionBarColumnWidth - pageActionInBreadcrumbMinWidth, "px");
        }
      }
    }

    utilSetCustomCSSProps(headerRef, (_utilSetCustomCSSProp = {}, _defineProperty(_utilSetCustomCSSProp, "--".concat(blockClass, "--max-action-bar-width-px"), newActionBarWidth), _defineProperty(_utilSetCustomCSSProp, "--".concat(blockClass, "--button-set-in-breadcrumb-width-px"), "".concat(newPageActionInBreadcrumbWidth)), _utilSetCustomCSSProp));
  }, [actionBarColumnWidth, actionBarMaxWidth, actionBarMinWidth, pageActionInBreadcrumbMaxWidth, pageActionInBreadcrumbMinWidth, headerRef]);
  useEffect(function () {
    var _utilSetCustomCSSProp2;

    // Updates custom CSS props used to manage scroll behaviour
    utilSetCustomCSSProps(headerRef, (_utilSetCustomCSSProp2 = {}, _defineProperty(_utilSetCustomCSSProp2, "--".concat(blockClass, "--height-px"), "".concat(metrics.headerHeight, "px")), _defineProperty(_utilSetCustomCSSProp2, "--".concat(blockClass, "--width-px"), "".concat(metrics.headerWidth, "px")), _defineProperty(_utilSetCustomCSSProp2, "--".concat(blockClass, "--header-top"), "".concat(metrics.headerTopValue + pageHeaderOffset, "px")), _defineProperty(_utilSetCustomCSSProp2, "--".concat(blockClass, "--breadcrumb-title-visibility"), scrollYValue > 0 ? 'visible' : 'hidden'), _defineProperty(_utilSetCustomCSSProp2, "--".concat(blockClass, "--scroll"), "".concat(scrollYValue)), _defineProperty(_utilSetCustomCSSProp2, "--".concat(blockClass, "--breadcrumb-title-top"), "".concat(Math.max(0, metrics.breadcrumbTitleHeight + metrics.titleRowSpaceAbove - scrollYValue), "px")), _defineProperty(_utilSetCustomCSSProp2, "--".concat(blockClass, "--breadcrumb-title-opacity"), "".concat(Math.min(1, Math.max(0, (scrollYValue - (metrics.titleRowSpaceAbove || 0)) / (metrics.breadcrumbTitleHeight || 1) // don't want to divide by zero
    )))), _defineProperty(_utilSetCustomCSSProp2, "--".concat(blockClass, "--breadcrumb-row-width-px"), "".concat(metrics.breadcrumbRowWidth, "px")), _utilSetCustomCSSProp2));
  }, [headerRef, preventBreadcrumbScroll, metrics, metrics.breadcrumbRowHeight, metrics.breadcrumbRowSpaceBelow, metrics.breadcrumbTitleHeight, metrics.breadcrumbRowWidth, metrics.headerHeight, metrics.headerWidth, metrics.headerTopValue, metrics.navigationRowHeight, navigation, pageHeaderOffset, scrollYValue, tags]);
  useWindowScroll( // on scroll or various layout changes check updates if needed
  function (_ref4) {
    var current = _ref4.current;
    utilSetCustomCSSProps(headerRef, _defineProperty({}, "--".concat(blockClass, "--breadcrumb-top"), "".concat(pageHeaderOffset, "px")));
    var fullyCollapsed = current.scrollY + metrics.headerTopValue + pageHeaderOffset >= 0;
    setFullyCollapsed(fullyCollapsed); // set offset for tagset tooltip

    var tagsetTooltipOffset = fullyCollapsed ? metrics.headerHeight + metrics.headerTopValue + pageHeaderOffset : metrics.headerHeight + pageHeaderOffset;
    document.documentElement.style.setProperty("--".concat(blockClass, "--tagset-tooltip-position"), fullyCollapsed ? 'fixed' : 'absolute');
    document.documentElement.style.setProperty("--".concat(blockClass, "--tagset-tooltip-offset"), "".concat(tagsetTooltipOffset, "px"));
    setScrollYValue(current.scrollY);
  }, [metrics.headerHeight, metrics.headerTopValue, pageHeaderOffset]);
  useWindowResize(function () {
    // on window resieze and other updates some values may have changed
    checkUpdateVerticalSpace();
  }, [actionBarItems, availableSpace, breadcrumbItems, preventBreadcrumbScroll, navigation, pageActions, subtitle, tags, title]);
  useEffect(function () {
    // Determines if the background should be one based on the header height or scroll
    var result = background && 1;

    if (!result && metrics.headerHeight > 0 && (breadcrumbItems || actionBarItems || tags || navigation)) {
      var startAddingAt = parseFloat(layout05, 10) * parseInt(baseFontSize);
      var scrollRemaining = metrics.headerHeight - scrollYValue;
      /* don't know how to test resize */

      /* istanbul ignore next if */

      if (scrollRemaining < startAddingAt) {
        var distanceAddingOver = startAddingAt - metrics.breadcrumbRowHeight;
        result = Math.min(1, (startAddingAt - scrollRemaining) / distanceAddingOver);
      }
    }

    utilSetCustomCSSProps(headerRef, _defineProperty({}, "--".concat(blockClass, "--background-opacity"), result));
    setBackgroundOpacity(result);
    setHasCollapseButton(collapseHeaderToggleWanted && result > 0);
  }, [actionBarItems, background, breadcrumbItems, headerRef, metrics.breadcrumbRowHeight, metrics.headerHeight, navigation, scrollYValue, collapseHeaderToggleWanted, tags]);
  useEffect(function () {
    // Determine if space is needed in the breadcrumb for a collapse button
    setSpaceForCollapseButton(hasCollapseButton && !(navigation || tags) && metrics.headerHeight);
  }, [hasCollapseButton, navigation, tags, metrics.headerHeight]);

  var nextToTabsCheck = function nextToTabsCheck() {
    return preventBreadcrumbScroll && actionBarItems === undefined && scrollYValue + metrics.headerTopValue >= 0;
  };

  useEffect(function () {
    toggleCollapse(collapseHeader); // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [collapseHeader]);
  var titleText = titleShape.text,
      TitleIcon = titleShape.icon,
      titleLoading = titleShape.loading;
  var tempTagSetOverflowProps = {
    // temp fix pending changes to PageHeader implementation
    allTagsModalTitle: 'All tags',
    allTagsModalSearchLabel: 'Search all tags',
    allTagsModalSearchPlaceholderText: 'Search all tags',
    showAllTagsLabel: 'View all tags'
  };
  return /*#__PURE__*/React.createElement(ReactResizeDetector, {
    handleHeight: true,
    onResize: handleResize
  }, /*#__PURE__*/React.createElement("section", _extends({}, rest, {
    className: cx([blockClass, "".concat(blockClass, "--no-margins-below-row"), className, (_ref5 = {}, _defineProperty(_ref5, "".concat(blockClass, "--show-background"), backgroundOpacity > 0), _defineProperty(_ref5, "".concat(blockClass, "--has-navigation"), navigation || tags), _defineProperty(_ref5, "".concat(blockClass, "--has-navigation-tags-only"), !navigation && tags), _ref5)]),
    ref: headerRef
  }), /*#__PURE__*/React.createElement(Grid, null, /*#__PURE__*/React.createElement("div", {
    className: "".concat(blockClass, "__non-navigation-row-contnet")
  }, hasBreadcrumbRow ? /*#__PURE__*/React.createElement(Row, {
    className: cx("".concat(blockClass, "__breadcrumb-row"), (_cx = {}, _defineProperty(_cx, "".concat(blockClass, "__breadcrumb-row--next-to-tabs"), nextToTabsCheck()), _defineProperty(_cx, "".concat(blockClass, "__breadcrumb-row--has-breadcrumbs"), breadcrumbItems), _defineProperty(_cx, "".concat(blockClass, "__breadcrumb-row--has-action-bar"), hasActionBar), _cx))
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(blockClass, "__breadcrumb-row--container")
  }, /*#__PURE__*/React.createElement(Column, {
    className: cx("".concat(blockClass, "__breadcrumb-column"), _defineProperty({}, "".concat(blockClass, "__breadcrumb-column--background"), breadcrumbItems !== undefined || hasActionBar))
  }, breadcrumbItems !== undefined ? /*#__PURE__*/React.createElement(BreadcrumbWithOverflow, {
    className: "".concat(blockClass, "__breadcrumb"),
    noTrailingSlash: title !== undefined
  }, breadcrumbItems, title ? /*#__PURE__*/React.createElement(BreadcrumbItem, {
    href: "#",
    isCurrentPage: true,
    className: cx(["".concat(blockClass, "__breadcrumb-title"), _defineProperty({}, "".concat(blockClass, "__breadcrumb-title--pre-collapsed"), preCollapseTitleRow)])
  }, titleLoading ? /*#__PURE__*/React.createElement(SkeletonText, null) : titleText) : '') : ''), /*#__PURE__*/React.createElement(Column, {
    className: cx(["".concat(blockClass, "__action-bar-column ").concat(blockClass, "__action-bar-column--background"), (_ref7 = {}, _defineProperty(_ref7, "".concat(blockClass, "__action-bar-column--has-page-actions"), pageActions), _defineProperty(_ref7, "".concat(blockClass, "__action-bar-column--influenced-by-collapse-button"), spaceForCollapseButton), _ref7)])
  }, /*#__PURE__*/React.createElement(ReactResizeDetector, {
    handleWidth: true,
    onResize: handleResizeActionBarColumn
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(blockClass, "__action-bar-column-content")
  }, hasActionBar ?
  /*#__PURE__*/
  // Investigate the responsive  behaviour or this and the title also fix the ActionBar Item and PageAction story css
  React.createElement(React.Fragment, null, pageActions && /*#__PURE__*/React.createElement("div", {
    className: cx("".concat(blockClass, "__page-actions"), _defineProperty({}, "".concat(blockClass, "__page-actions--in-breadcrumb"), pageActionsInBreadcrumbRow))
  }, /*#__PURE__*/React.createElement(ButtonSetWithOverflow, {
    className: "".concat(blockClass, "__button-set--in-breadcrumb"),
    onWidthChange: handleButtonSetWidthChange,
    buttons: pageActionsItemArray
  })), /*#__PURE__*/React.createElement(ActionBar, {
    actions: actionBarItemArray,
    className: "".concat(blockClass, "__action-bar"),
    onWidthChange: handleActionBarWidthChange,
    rightAlign: true
  })) : null))))) : null, !preCollapseTitleRow && !(title === undefined && pageActions === undefined) ? /*#__PURE__*/React.createElement(Row, {
    className: cx("".concat(blockClass, "__title-row"), (_cx4 = {}, _defineProperty(_cx4, "".concat(blockClass, "__title-row--no-breadcrumb-row"), !hasBreadcrumbRow), _defineProperty(_cx4, "".concat(blockClass, "__title-row--under-action-bar"), hasActionBar), _defineProperty(_cx4, "".concat(blockClass, "__title-row--has-page-actions"), pageActions !== undefined), _defineProperty(_cx4, "".concat(blockClass, "__title-row--sticky"), pageActions !== undefined && actionBarItems === undefined && hasBreadcrumbRow), _cx4))
  }, /*#__PURE__*/React.createElement(Column, {
    className: "".concat(blockClass, "__title-column")
  }, title !== undefined ? /*#__PURE__*/React.createElement("div", {
    className: cx("".concat(blockClass, "__title"), _defineProperty({}, "".concat(blockClass, "__title--fades"), hasBreadcrumbRow))
  }, TitleIcon && !titleLoading ? /*#__PURE__*/React.createElement(TitleIcon, {
    className: "".concat(blockClass, "__title-icon")
  }) : null, /*#__PURE__*/React.createElement("span", {
    title: !titleLoading ? titleText : null
  }, titleLoading ? /*#__PURE__*/React.createElement(SkeletonText, {
    className: "".concat(blockClass, "__title-skeleton")
  }) : titleText)) : null), pageActions !== undefined ? /*#__PURE__*/React.createElement(Column, {
    className: cx("".concat(blockClass, "__page-actions"), _defineProperty({}, "".concat(blockClass, "__page-actions--in-breadcrumb"), pageActionsInBreadcrumbRow))
  }, /*#__PURE__*/React.createElement(ButtonSetWithOverflow, {
    className: "".concat(blockClass, "__page-actions-container"),
    onWidthChange: handleButtonSetWidthChange,
    buttons: pageActionsItemArray
  })) : null) : null, subtitle !== undefined ? /*#__PURE__*/React.createElement(Row, {
    className: "".concat(blockClass, "__subtitle-row")
  }, /*#__PURE__*/React.createElement(Column, {
    className: "".concat(blockClass, "__subtitle")
  }, subtitle)) : null, availableSpace !== undefined ? /*#__PURE__*/React.createElement(Row, {
    className: "".concat(blockClass, "__available-row")
  }, /*#__PURE__*/React.createElement(Column, {
    className: "".concat(blockClass, "__available-column")
  }, availableSpace)) : null, (breadcrumbItems || actionBarItems || title || pageActions || availableSpace || subtitle) && /*#__PURE__*/React.createElement("div", {
    className: cx(["".concat(blockClass, "__last-row-buffer"), _defineProperty({}, "".concat(blockClass, "__last-row-buffer--active"), lastRowBufferActive)])
  }), // this navigation row scrolls under the breadcrumb if there is one
  tags && !navigation ? /*#__PURE__*/React.createElement(Row, {
    className: cx("".concat(blockClass, "__navigation-row"), _defineProperty({}, "".concat(blockClass, "__navigation-row--has-tags"), tags))
  }, /*#__PURE__*/React.createElement(Column, {
    className: cx("".concat(blockClass, "__navigation-tags"), _defineProperty({}, "".concat(blockClass, "__navigation-tags--tags-only"), navigation === undefined))
  }, /*#__PURE__*/React.createElement(TagSet, _extends({
    overflowAlign: "end"
  }, tempTagSetOverflowProps, {
    tags: tags
  })))) : null), // this navigation pushes the breadcrumb off or settles underneath it depending on preventBreadcrumbScroll
  navigation ? /*#__PURE__*/React.createElement(Row, {
    className: cx("".concat(blockClass, "__navigation-row"), (_cx9 = {}, _defineProperty(_cx9, "".concat(blockClass, "__navigation-row--spacing-above-06"), navigation !== undefined), _defineProperty(_cx9, "".concat(blockClass, "__navigation-row--has-tags"), tags), _cx9))
  }, /*#__PURE__*/React.createElement(Column, {
    className: "".concat(blockClass, "__navigation-tabs")
  }, navigation), tags !== undefined ? /*#__PURE__*/React.createElement(Column, {
    className: cx("".concat(blockClass, "__navigation-tags"), _defineProperty({}, "".concat(blockClass, "__navigation-tags--tags-only"), navigation === undefined))
  }, /*#__PURE__*/React.createElement(TagSet, _extends({
    overflowAlign: "end"
  }, tempTagSetOverflowProps, {
    tags: tags
  }))) : null) : null), hasCollapseButton ? /*#__PURE__*/React.createElement(Button, {
    className: cx("".concat(blockClass, "__collapse-expand-toggle"), _defineProperty({}, "".concat(blockClass, "__collapse-expand-toggle--collapsed"), fullyCollapsed)),
    "data-collapse": fullyCollapsed ? 'collapsed' : 'not collapsed',
    hasIconOnly: true,
    iconDescription: fullyCollapsed ? expandHeaderLabel : collapseHeaderLabel,
    kind: "ghost",
    onClick: handleCollapseToggle,
    renderIcon: ChevronUp16,
    size: "field",
    tooltipPosition: "bottom",
    tooltipAlignment: "end",
    type: "button"
  }) : null));
}); // Return a placeholder if not released and not enabled by feature flag

PageHeader = pkg.checkComponentEnabled(PageHeader, componentName); // copied from carbon-components-react/src/components/Tag/Tag.js for DocGen

var TYPES = {
  red: 'Red',
  magenta: 'Magenta',
  purple: 'Purple',
  blue: 'Blue',
  cyan: 'Cyan',
  teal: 'Teal',
  green: 'Green',
  gray: 'Gray',
  'cool-gray': 'Cool-Gray',
  'warm-gray': 'Warm-Gray',
  'high-contrast': 'High-Contrast'
};
var tagTypes = Object.keys(TYPES);
PageHeader.propTypes = {
  /**
   * Specifies the action bar items. Each item is specified as an object
   * with the properties of a Carbon Button in icon only form. Button kind, size, tooltipPosition,
   * tooltipAlignment and type are ignored.
   */
  actionBarItems: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.shape(_objectSpread(_objectSpread({}, prepareProps(Button.propTypes, ['kind', 'size', 'tooltipPosition', 'tooltipAlignment'])), {}, {
    iconDescription: PropTypes.string.isRequired,
    onClick: Button.propTypes.onClick,
    renderIcon: PropTypes.oneOfType([PropTypes.func, PropTypes.object]).isRequired
  }))), deprecatePropUsage(PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.element), PropTypes.element]), 'Expects an array of objects with the following properties: iconDescription, renderIcon and onClick.')]),
  // expects action bar item as array or in fragment

  /**
   * A zone for placing high-level, client content above the page tabs.
   * Accepts arbitrary renderable content as a React node. Optional.
   */
  availableSpace: PropTypes.node,

  /**
   * Specifies if the PageHeader should have a background color or not.
   * Optional.
   */
  background: PropTypes.bool,

  /**
   * One or more Carbon BreadcrumbItem components, passed in as React element(s).
   * If provided, these are rendered at the top left of the header as
   * a breadcrumb. Optional.
   */
  breadcrumbItems: PropTypes.element,
  // expects BreadcrumbItems,

  /**
   * Specifies class(es) to be applied to the top-level PageHeader node.
   * Optional.
   */
  className: PropTypes.string,

  /**
   * The header can as a whole be collapsed, expanded or somewhere in between.
   * This setting controls the initial value, but also takes effect on change
   *
   * NOTE: The header is collapsed by setting the scroll position to hide part of the header. Collapsing has no effect if there is insufficient content to scroll.
   */
  collapseHeader: PropTypes.bool,

  /**
   * Label/assistive text for the collapse/expand button
   * Default 'Collapse'
   */
  collapseHeaderLabel: PropTypes.string,

  /**
   * Enable the collapse header toggle.
   *
   * NOTE: The header is collapsed by setting the scroll position to hide part of the header. Collapsing has no effect if there is insufficient content to scroll.
   */
  collapseHeaderToggleWanted: PropTypes.bool,

  /**
   * Label/assistive text for the collapse/expand button
   * Default 'Expand'
   */
  expandHeaderLabel: PropTypes.string,

  /**
   * Content for the navigation area in the PageHeader. Should
   * be a React element that is normally a Carbon Tabs component. Optional.
   */
  navigation: PropTypes.element,
  // Supports Tabs

  /**
   * Specifies the primary page actions. Each action is specified as an object
   * with the properties of a Carbon Button plus:
   * - label: node
   *
   * Carbon Button API https://react.carbondesignsystem.com/?path=/docs/components-button--default#component-api
   */
  pageActions: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.shape(_objectSpread(_objectSpread({}, Button.propTypes), {}, {
    kind: Button.propTypes.kind,
    label: PropTypes.node.isRequired,
    onClick: Button.propTypes.onClick
  }))), deprecatePropUsage(PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.element), PropTypes.element]), 'Expects an array of objects with the following properties: label and onClick.')]),

  /**
   * Number of pixels the page header sits from the top of the screen.
   * The nature of the pageHeader makes this hard to measure
   */
  pageHeaderOffset: PropTypes.number,

  /**
   * The title row typically starts below the breadcrumb row. This option
   * preCollapses it into the breadcrumb row.
   */
  preCollapseTitleRow: PropTypes.bool,

  /**
   * Standard behavior scrolls the breadcrumb off to leave just tabs. This
   * option preserves vertical space for both.
   */
  preventBreadcrumbScroll: PropTypes.bool,

  /**
   * A subtitle or description that provides additional context to
   * identify the current page. Optional.
   */
  subtitle: PropTypes.string,

  /**
   * The tags to be shown in the PageHeader. Each tag is specified as an object
   * with properties: **label**\* (required) to supply the tag content, and
   * other properties will be passed to the Carbon Tag component, such as
   * **type**, **disabled**, **ref**, **className** , and any other Tag props.
   * NOTE: **filter** is not supported. Any other fields in the object will be passed through to the HTML element
   * as HTML attributes.
   *
   * See https://react.carbondesignsystem.com/?path=/docs/components-tag--default
   */
  tags: PropTypes.arrayOf(PropTypes.shape(_objectSpread(_objectSpread({}, prepareProps(Tag.propTypes, 'filter')), {}, {
    label: PropTypes.string.isRequired,
    // we duplicate this prop to improve the DocGen
    type: PropTypes.oneOf(tagTypes)
  }))),

  /**
   * The title of the page.
   * Optional string or object with the following attributes: text, icon, loading
   */
  title: PropTypes.oneOfType([PropTypes.shape({
    text: PropTypes.string.isRequired,
    icon: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    loading: PropTypes.bool
  }), PropTypes.string]),

  /**
   * An icon to be included to the left of the title text.
   * Optional.
   */
  titleIcon: deprecateProp(PropTypes.oneOfType([PropTypes.func, PropTypes.object]), 'Use `title` prop shape instead.')
};
PageHeader.defaultProps = {
  background: true,
  className: '',
  collapseHeaderLabel: 'Collapse',
  expandHeaderLabel: 'Expand',
  preventBreadcrumbScroll: false,
  pageHeaderOffset: 0,
  preCollapseTitleRow: false
};
PageHeader.displayName = componentName;
