// @carbon/ibm-cloud-cognitive/es/components/PageHeader/PageHeaderUtils.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import { pkg } from '../../settings';
export var blockClass = "".concat(pkg.prefix, "--page-header");
/**
 * Assesses the vertical height of various elements and calls setMetrics with update
 * @param {{}} headerRef
 * @param {{}} navigation
 * @param {boolean} preventBreadcrumbScroll
 * @param {()} setMetrics
 */

export var utilCheckUpdateVerticalSpace = function utilCheckUpdateVerticalSpace(headerRef, navigation, preventBreadcrumbScroll, setMetrics) {
  var dynamicRefs = {};

  var getDynamicRef = function getDynamicRef(selector) {
    // would love to do this differently but digging in the dom seems easier
    // than getting a ref to a conditionally rendered item

    /* don't know how to test resize */

    /* istanbul ignore next if */
    if (!headerRef.current) {
      return undefined;
    } else {
      var dRef = dynamicRefs[selector];

      if (!dRef || dRef.parentNode === null) {
        dynamicRefs[selector] = headerRef.current.querySelector(selector);
      }
    }

    return dynamicRefs[selector];
  }; // Utility function that checks the heights of various elements which are used to determine layout


  var update = {};
  var breadcrumbTitleEl = getDynamicRef(".".concat(blockClass, "__breadcrumb-title"));
  var breadcrumbRowEl = getDynamicRef(".".concat(blockClass, "__breadcrumb-row"));
  var titleRowEl = getDynamicRef(".".concat(blockClass, "__title-row"));
  var subtitleRowEl = getDynamicRef(".".concat(blockClass, "__subtitle-row"));
  var availableRowEl = getDynamicRef(".".concat(blockClass, "__available-row"));
  var navigationRowEl = getDynamicRef(".".concat(blockClass, "__navigation-row"));
  update.headerHeight = headerRef.current ? headerRef.current.clientHeight : 0;
  update.headerWidth = headerRef.current ? headerRef.current.offsetWidth : 0;
  update.breadcrumbRowHeight = breadcrumbRowEl ? breadcrumbRowEl.clientHeight : 0;
  update.breadcrumbRowWidth = breadcrumbRowEl ? breadcrumbRowEl.offsetWidth : 0;
  update.breadcrumbTitleHeight = breadcrumbTitleEl ? breadcrumbTitleEl.clientHeight : 1;
  update.titleRowHeight = titleRowEl ? titleRowEl.clientHeight : 0;
  update.subtitleRowHeight = subtitleRowEl ? subtitleRowEl.clientHeight : 0;
  update.availableRowHeight = availableRowEl ? availableRowEl.clientHeight : 0;
  update.navigationRowHeight = navigationRowEl ? navigationRowEl.clientHeight : 1;
  update.breadcrumbRowSpaceBelow = 0;
  update.titleRowSpaceAbove = 0; // Base for calculating sticky top

  update.headerTopValue = -update.headerHeight;

  if (navigation) {
    // adjust top for sticky with navigation
    update.headerTopValue += update.navigationRowHeight;
  }

  if (preventBreadcrumbScroll || !navigation) {
    // adjust sticky top if no navigation or breadcrumb is to stay on screen
    update.headerTopValue += update.breadcrumbRowHeight;
  }

  if (window) {
    var val;
    /* don't know how to test resize */

    /* istanbul ignore next if */

    if (breadcrumbRowEl) {
      val = parseFloat(window.getComputedStyle(breadcrumbRowEl).getPropertyValue('margin-bottom'), 10);
      update.breadcrumbRowSpaceBelow = isNaN(val) ? 0 : val;
    }
    /* don't know how to test resize */

    /* istanbul ignore next if */


    if (titleRowEl) {
      val = parseFloat(window.getComputedStyle(titleRowEl).getPropertyValue('margin-top'), 10);
      update.titleRowSpaceAbove = isNaN(val) ? 0 : val;
    }
  }

  setMetrics(function (previous) {
    return _objectSpread(_objectSpread({}, previous), update);
  });
};
/**
 * Takes the title parameters and returns a titleShape
 * @param {{} | string} title
 * @param {{}} titleIcon
 * @param {string} defaultTitle
 * @returns
 */

export var utilGetTitleShape = function utilGetTitleShape(title, titleIcon, defaultTitle) {
  // Title shape is used to allow title to be string or shape
  var newShape = _objectSpread({}, defaultTitle);

  if (title) {
    if (title.text) {
      // title is in shape format
      newShape = Object.assign(newShape, _objectSpread({}, title));
    } else {
      // title is a string
      newShape.text = title;
    }
  }

  if (!newShape.icon && titleIcon) {
    // if no icon use titleIcon if supplied
    newShape.icon = titleIcon;
  }

  return newShape;
}; // Set css style values directly onto a node. Note that this is effective
// immediately, rather than using the React `style` prop which goes through
// the React DOM update optimisation.

export var utilSetCustomCSSProps = function utilSetCustomCSSProps(targetRef, kvPairs) {
  if (targetRef.current) {
    var keys = Object.keys(kvPairs);

    for (var _i = 0, _keys = keys; _i < _keys.length; _i++) {
      var k = _keys[_i];
      targetRef.current.style.setProperty(k, kvPairs[k]);
    }
  }
};
