// @carbon/ibm-cloud-cognitive/lib/components/CreateSidePanel/CreateSidePanel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CreateSidePanel = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _settings = require("../../settings");

var _propsHelper = require("../../global/js/utils/props-helper");

var _carbonComponentsReact = require("carbon-components-react");

var _SidePanel = require("../SidePanel/SidePanel");

var _excluded = ["className", "onRequestClose", "open", "title", "subtitle", "children", "disableSubmit", "onRequestSubmit", "primaryButtonText", "secondaryButtonText", "selectorPrimaryFocus", "pageContentSelector", "selectorPageContent", "formTitle", "formDescription"];
// The block part of our conventional BEM class names (blockClass__E--M).
var blockClass = "".concat(_settings.pkg.prefix, "--create-side-panel");
var componentName = 'CreateSidePanel'; // NOTE: the component SCSS is not imported here: it is rolled up separately.

/**
 * This is an example component to show relevant conventions and usage.
 */

var CreateSidePanel = /*#__PURE__*/_react.default.forwardRef(function (_ref, ref) {
  var className = _ref.className,
      onRequestClose = _ref.onRequestClose,
      open = _ref.open,
      title = _ref.title,
      subtitle = _ref.subtitle,
      children = _ref.children,
      disableSubmit = _ref.disableSubmit,
      onRequestSubmit = _ref.onRequestSubmit,
      primaryButtonText = _ref.primaryButtonText,
      secondaryButtonText = _ref.secondaryButtonText,
      selectorPrimaryFocus = _ref.selectorPrimaryFocus,
      pageContentSelector = _ref.pageContentSelector,
      selectorPageContent = _ref.selectorPageContent,
      formTitle = _ref.formTitle,
      formDescription = _ref.formDescription,
      rest = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  var actions = [{
    label: primaryButtonText,
    onClick: onRequestSubmit,
    kind: 'primary',
    disabled: disableSubmit
  }, {
    label: secondaryButtonText,
    onClick: onRequestClose,
    kind: 'secondary'
  }];
  return (selectorPageContent || pageContentSelector) && /*#__PURE__*/_react.default.createElement(_SidePanel.SidePanel, (0, _extends2.default)({}, rest, {
    open: open,
    ref: ref,
    pageContentSelector: pageContentSelector,
    selectorPageContent: selectorPageContent,
    onRequestClose: onRequestClose,
    title: title,
    subtitle: subtitle,
    actions: actions,
    selectorPrimaryFocus: selectorPrimaryFocus,
    placement: "right",
    slideIn: true,
    animateTitle: false,
    className: (0, _classnames.default)(blockClass, className),
    size: "md"
  }), /*#__PURE__*/_react.default.createElement("h3", {
    className: "".concat(blockClass, "__form-title-text ").concat(blockClass, "__content-text")
  }, formTitle), /*#__PURE__*/_react.default.createElement("p", {
    className: "".concat(blockClass, "__form-description-text ").concat(blockClass, "__content-text")
  }, formDescription), /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.Form, {
    className: "".concat(blockClass, "__form")
  }, children));
});

exports.CreateSidePanel = CreateSidePanel;
exports.CreateSidePanel = CreateSidePanel = _settings.pkg.checkComponentEnabled(CreateSidePanel, componentName);
CreateSidePanel.displayName = componentName;
CreateSidePanel.propTypes = {
  /**
   * Sets the body content of the create side panel
   */
  children: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.node), _propTypes.default.node]).isRequired,

  /**
   * Provide an optional class to be applied to the containing node.
   */
  className: _propTypes.default.string,

  /**
   * The description of the CreateSidePanel serves to provide more information about the form within the panel.
   */
  description: _propTypes.default.node,

  /**
   * Specifies a boolean for disabling or enabling the primary button. This is important for form validation
   * Returning `true` prevents the primary button from being clicked until required fields are completed.
   */
  disableSubmit: _propTypes.default.bool,

  /**
   * Specifies an optional field that provides a additional context for a form
   */
  formDescription: _propTypes.default.node,

  /**
   * Specifies a required field that provides a title for a form
   */
  formTitle: _propTypes.default.node.isRequired,

  /**
   * Specifies an optional handler which is called when the CreateSidePanel
   * is closed.
   */
  onRequestClose: _propTypes.default.func,

  /**
   * Specifies an optional handler which is called when the CreateSidePanel
   * primary button is pressed.
   */
  onRequestSubmit: _propTypes.default.func,

  /**
   * Specifies whether the CreateSidePanel is open or not.
   */
  open: _propTypes.default.bool,

  /**
   * This is the selector to the element that contains all of the page content that will shrink if the panel is a slide in.
   * This prop is required when using the `slideIn` variant of the side panel.
   */
  pageContentSelector: (0, _propsHelper.deprecateProp)(_propTypes.default.string, 'This prop has been renamed to `selectorPageContent`.'),

  /**
   * Specifies the primary button's text in the modal.
   */
  primaryButtonText: _propTypes.default.string.isRequired,

  /**
   * Specifies the secondary button's text in the modal.
   */
  secondaryButtonText: _propTypes.default.string.isRequired,

  /**
   * This is the selector to the element that contains all of the page content that will shrink if the panel is a slide in.
   * This prop is required when using the `slideIn` variant of the side panel.
   */
  selectorPageContent: _propTypes.default.string.isRequired,

  /**
   * Specifies which DOM element in the form should be focused.
   */
  selectorPrimaryFocus: _propTypes.default.node.isRequired,

  /**
   * The subtitle of the CreateSidePanel is optional and serves to provide more information about the modal.
   */
  subtitle: _propTypes.default.node,

  /**
   * The title of the CreateSidePanel is usually the product or service name.
   */
  title: _propTypes.default.node.isRequired
};
