// @carbon/ibm-cloud-cognitive/lib/components/ButtonSetWithOverflow/ButtonSetWithOverflow.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ButtonSetWithOverflow = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _reactResizeDetector = _interopRequireDefault(require("react-resize-detector"));

var _carbonComponentsReact = require("carbon-components-react");

var _ButtonMenu = require("../ButtonMenu");

var _settings = require("../../settings");

var _propsHelper = require("../../global/js/utils/props-helper");

var _excluded = ["buttons", "size"],
    _excluded2 = ["label"],
    _excluded3 = ["buttons"],
    _excluded4 = ["label", "kind"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

var blockClass = "".concat(_settings.pkg.prefix, "--button-set-with-overflow");
var componentName = 'ButtonSetWithOverflow';

var ButtonSetWithOverflow = function ButtonSetWithOverflow(_ref) {
  var buttons = _ref.buttons,
      children = _ref.children,
      className = _ref.className,
      onWidthChange = _ref.onWidthChange,
      pageActionsLabel = _ref.pageActionsLabel,
      rightAlign = _ref.rightAlign,
      size = _ref.size;

  var _useState = (0, _react.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      showAsOverflow = _useState2[0],
      setShowAsOverflow = _useState2[1];

  var spaceAvailableRef = (0, _react.useRef)(null);
  var sizingContainerRefSet = (0, _react.useRef)(null);
  var sizingContainerRefCombo = (0, _react.useRef)(null);

  var _useState3 = (0, _react.useState)([]),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      itemArray = _useState4[0],
      setItemArray = _useState4[1]; // determine display count based on space available and width of pageActions


  var checkFullyVisibleItems = function checkFullyVisibleItems() {
    var _spaceAvailableRef$cu, _sizingContainerRefSe, _sizingContainerRefCo, _sizingContainerRefCo2;

    var spaceAvailable = (_spaceAvailableRef$cu = spaceAvailableRef.current) === null || _spaceAvailableRef$cu === void 0 ? void 0 : _spaceAvailableRef$cu.offsetWidth;
    var newShowAsOverflow = true;
    var sizingSet = (_sizingContainerRefSe = sizingContainerRefSet.current) === null || _sizingContainerRefSe === void 0 ? void 0 : _sizingContainerRefSe.querySelectorAll(".".concat(_settings.carbon.prefix, "--btn"));
    var sizingSetTotalSize = 0;

    var _iterator = _createForOfIteratorHelper(sizingSet),
        _step;

    try {
      for (_iterator.s(); !(_step = _iterator.n()).done;) {
        var item = _step.value;
        sizingSetTotalSize += item.offsetWidth;
      }
    } catch (err) {
      _iterator.e(err);
    } finally {
      _iterator.f();
    }

    var sizingComboSize = (_sizingContainerRefCo = sizingContainerRefCombo.current) !== null && _sizingContainerRefCo !== void 0 && _sizingContainerRefCo.offsetWidth ? (_sizingContainerRefCo2 = sizingContainerRefCombo.current) === null || _sizingContainerRefCo2 === void 0 ? void 0 : _sizingContainerRefCo2.offsetWidth : 0;
    onWidthChange && onWidthChange({
      maxWidth: sizingSetTotalSize,
      minWidth: sizingComboSize
    });

    if (sizingSetTotalSize <= spaceAvailable) {
      newShowAsOverflow = false;
    }

    setShowAsOverflow(newShowAsOverflow);
  };

  (0, _react.useEffect)(function () {
    checkFullyVisibleItems(); // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [itemArray]);

  var handleResize = function handleResize() {
    // width is the space available for all action bar items horizontally
    // the action bar items are squares so the height should be one item wide

    /* istanbul ignore next */
    // not sure how to test resize
    checkFullyVisibleItems();
  };

  var handleButtonResize = function handleButtonResize() {
    /* istanbul ignore next */
    // not sure how to test resize
    checkFullyVisibleItems();
  };

  (0, _react.useEffect)(function () {
    if (buttons) {
      setItemArray(buttons);
    } else {
      var _extractShapesArray;

      setItemArray((_extractShapesArray = (0, _propsHelper.extractShapesArray)(children)) === null || _extractShapesArray === void 0 ? void 0 : _extractShapesArray.map(function (shape) {
        return _objectSpread({
          label: shape.children
        }, shape);
      }));
    }
  }, [buttons, children]);

  var AButtonSet = /*#__PURE__*/_react.default.forwardRef(function (_ref2, ref) {
    var buttons = _ref2.buttons,
        size = _ref2.size,
        rest = (0, _objectWithoutProperties2.default)(_ref2, _excluded);
    return /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.ButtonSet, (0, _extends2.default)({}, rest, {
      ref: ref
    }), buttons.map(function (_ref3, index) {
      var label = _ref3.label,
          other = (0, _objectWithoutProperties2.default)(_ref3, _excluded2);
      return /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.Button, (0, _extends2.default)({
        key: index
      }, other, {
        size: size,
        type: "button"
      }), label);
    }));
  });

  var AButtonMenu = /*#__PURE__*/_react.default.forwardRef(function (_ref4, ref) {
    var buttons = _ref4.buttons,
        rest = (0, _objectWithoutProperties2.default)(_ref4, _excluded3);
    return /*#__PURE__*/_react.default.createElement(_ButtonMenu.ButtonMenu, (0, _extends2.default)({}, rest, {
      ref: ref,
      label: pageActionsLabel
    }), buttons.map(function (_ref5, index) {
      var label = _ref5.label,
          kind = _ref5.kind,
          other = (0, _objectWithoutProperties2.default)(_ref5, _excluded4);
      return /*#__PURE__*/_react.default.createElement(_ButtonMenu.ButtonMenuItem, (0, _extends2.default)({
        key: index,
        isDelete: kind === 'danger',
        itemText: label
      }, (0, _propsHelper.prepareProps)(other, ['iconDescription', 'renderIcon'])));
    }).reverse());
  });

  return /*#__PURE__*/_react.default.createElement(_reactResizeDetector.default, {
    handleWidth: true,
    onResize: handleResize
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)([blockClass, className, (0, _defineProperty2.default)({}, "".concat(blockClass, "--right"), rightAlign)]),
    ref: spaceAvailableRef
  }, /*#__PURE__*/_react.default.createElement(_reactResizeDetector.default, {
    onResize: handleButtonResize
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(blockClass, "__button-container ").concat(blockClass, "__button-container--hidden")
  }, /*#__PURE__*/_react.default.createElement(AButtonSet, {
    "aria-hidden": true,
    ref: sizingContainerRefSet,
    size: size,
    buttons: itemArray
  }))), /*#__PURE__*/_react.default.createElement(_reactResizeDetector.default, {
    onResize: handleButtonResize
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(blockClass, "__button-container ").concat(blockClass, "__button-container--hidden"),
    "aria-hidden": true
  }, /*#__PURE__*/_react.default.createElement(AButtonMenu, {
    ref: sizingContainerRefCombo,
    buttons: itemArray,
    size: size
  }))), showAsOverflow ? /*#__PURE__*/_react.default.createElement(AButtonMenu, {
    buttons: itemArray,
    size: size
  }) : /*#__PURE__*/_react.default.createElement(AButtonSet, {
    className: "".concat(blockClass, "__button-container"),
    size: size,
    buttons: itemArray
  })));
};

exports.ButtonSetWithOverflow = ButtonSetWithOverflow;
ButtonSetWithOverflow.propTypes = {
  /**
  /**
   * Specifies the buttons for the ButtonSetWithOverflow. Each item is specified as an object
   * with the properties of a Carbon Button plus a label.
   *
   * Carbon Button API https://react.carbondesignsystem.com/?path=/docs/components-button--default#component-api
   */
  buttons: _propTypes.default.arrayOf(_propTypes.default.shape(_objectSpread(_objectSpread({}, _carbonComponentsReact.Button.propTypes), {}, {
    kind: _carbonComponentsReact.Button.propTypes.kind,
    label: _propTypes.default.node,
    onClick: _carbonComponentsReact.Button.propTypes.onClick
  }))),

  /**
   * children of the button set
   */
  children: (0, _propsHelper.deprecateProp)(_propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.element), _propTypes.default.element]), 'See documentation on the `buttons` property.'),
  // expects action bar item as array or in fragment,

  /**
   * className
   */
  className: _propTypes.default.string,

  /**
   * onResize reports maxSize on resize
   */
  onWidthChange: _propTypes.default.func,

  /**
   * pageActionsLabel - used when button set is shown as combo button
   */
  pageActionsLabel: _propTypes.default.node,

  /**
   * align buttons to right of available space
   */
  rightAlign: _propTypes.default.bool,

  /**
   * Specify the size of the button, from a list of available sizes.
   * For `default` buttons, this prop can remain unspecified.
   */
  size: _carbonComponentsReact.Button.propTypes.size
};
ButtonSetWithOverflow.defaultProps = {
  pageActionsLabel: 'Page actions',
  size: 'field'
};
ButtonSetWithOverflow.displayName = componentName;
