// @carbon/ibm-cloud-cognitive/es/components/Card/Card.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
var _excluded = ["id"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

//
// Copyright IBM Corp. 2020, 2021
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//
import React from 'react';
import cx from 'classnames';
import { Button, OverflowMenu, OverflowMenuItem } from 'carbon-components-react';
import PropTypes from 'prop-types';
import { pkg } from '../../settings';
var componentName = 'Card';
export var Card = function Card(_ref) {
  var _cx;

  var actionIcons = _ref.actionIcons,
      actionIconsPosition = _ref.actionIconsPosition,
      caption = _ref.caption,
      children = _ref.children,
      className = _ref.className,
      clickZone = _ref.clickZone,
      label = _ref.label,
      media = _ref.media,
      mediaPosition = _ref.mediaPosition,
      onClick = _ref.onClick,
      onKeyDown = _ref.onKeyDown,
      onPrimaryButtonClick = _ref.onPrimaryButtonClick,
      overflowActions = _ref.overflowActions,
      onSecondaryButtonClick = _ref.onSecondaryButtonClick,
      Pictogram = _ref.pictogram,
      primaryButtonKind = _ref.primaryButtonKind,
      primaryButtonText = _ref.primaryButtonText,
      productive = _ref.productive,
      secondaryButtonKind = _ref.secondaryButtonKind,
      secondaryButtonText = _ref.secondaryButtonText,
      title = _ref.title,
      titleSize = _ref.titleSize;
  var hasActions = actionIcons.length > 0 || overflowActions.length > 0;
  var hasTopActions = hasActions && actionIconsPosition === 'top';
  var hasBottomActions = hasActions && actionIconsPosition === 'bottom';
  var hasBottomButton = !!secondaryButtonText || !!primaryButtonText;
  var hasBottomBar = hasBottomActions || hasBottomButton;
  var headerClasses = cx("".concat(pkg.prefix, "-card-header"), (_cx = {}, _defineProperty(_cx, "".concat(pkg.prefix, "-card-header--label-only"), label && !title && !caption), _defineProperty(_cx, "".concat(pkg.prefix, "-card-header--has-label"), label && productive), _defineProperty(_cx, "".concat(pkg.prefix, "-card-title--lg"), titleSize === 'large'), _cx));
  var footerClasses = cx("".concat(pkg.prefix, "-card-footer"), _defineProperty({}, "".concat(pkg.prefix, "-card-footer--no-button"), !hasBottomButton));

  var getActions = function getActions() {
    if (overflowActions.length !== 0) {
      var pos = actionIconsPosition === 'top' ? 'bottom' : 'top';
      var size = actionIconsPosition === 'top' ? 'sm' : 'xl';
      return /*#__PURE__*/React.createElement(OverflowMenu, {
        size: size,
        direction: pos,
        flipped: true
      }, overflowActions.map(function (_ref2) {
        var id = _ref2.id,
            rest = _objectWithoutProperties(_ref2, _excluded);

        return /*#__PURE__*/React.createElement(OverflowMenuItem, _extends({
          key: id
        }, rest));
      }));
    }

    var icons = actionIcons.map(function (_ref3) {
      var id = _ref3.id,
          Icon = _ref3.icon,
          onClick = _ref3.onClick,
          iconDescription = _ref3.iconDescription,
          onKeyDown = _ref3.onKeyDown;

      if (productive) {
        return /*#__PURE__*/React.createElement(Button, {
          key: id,
          renderIcon: Icon,
          hasIconOnly: true,
          onClick: onClick,
          size: actionIconsPosition === 'top' ? 'sm' : 'field',
          iconDescription: iconDescription,
          kind: "ghost"
        });
      }

      return /*#__PURE__*/React.createElement("div", {
        key: id,
        className: "".concat(pkg.prefix, "-card-icon"),
        onClick: onClick,
        onKeyDown: onKeyDown,
        role: "button",
        tabIndex: "0"
      }, /*#__PURE__*/React.createElement(Icon, {
        "aria-label": iconDescription
      }));
    });
    return icons;
  };

  var getClickableProps = function getClickableProps() {
    return {
      onClick: onClick,
      onKeyDown: onKeyDown,
      role: 'button',
      tabIndex: '0'
    };
  };

  var getCardProps = function getCardProps() {
    var _cx3;

    var hasClickEvent = !!onClick || !!onKeyDown;
    var clickable = hasClickEvent && !productive || hasClickEvent && productive && clickZone === 'one';

    var cardProps = _objectSpread({
      className: cx("".concat(pkg.prefix, "-card"), (_cx3 = {}, _defineProperty(_cx3, "".concat(pkg.prefix, "-card--productive"), productive), _defineProperty(_cx3, "".concat(pkg.prefix, "-card--clickable"), clickable), _defineProperty(_cx3, "".concat(pkg.prefix, "-card--media-left"), mediaPosition === 'left'), _defineProperty(_cx3, "className", className), _cx3))
    }, clickable && getClickableProps());

    return cardProps;
  };

  var getHeaderBodyProps = function getHeaderBodyProps() {
    var hasClickEvent = !!onClick || !!onKeyDown;
    var clickable = hasClickEvent && clickZone === 'two';

    var headerBodyProps = _objectSpread({
      className: cx("".concat(pkg.prefix, "-card-header-body-container"), _defineProperty({}, "".concat(pkg.prefix, "-card--clickable"), clickable))
    }, clickable && getClickableProps());

    return headerBodyProps;
  };

  var getBodyProps = function getBodyProps() {
    var hasClickEvent = !!onClick || !!onKeyDown;
    var clickable = hasClickEvent && clickZone === 'three';

    var bodyProps = _objectSpread({
      className: cx("".concat(pkg.prefix, "-card-body"), _defineProperty({}, "".concat(pkg.prefix, "-card--clickable"), clickable))
    }, clickable && getClickableProps());

    return bodyProps;
  };

  var CardContent = /*#__PURE__*/React.createElement("div", getCardProps(), media && /*#__PURE__*/React.createElement("div", {
    className: "".concat(pkg.prefix, "-card-media")
  }, media), Pictogram && /*#__PURE__*/React.createElement("div", {
    className: "".concat(pkg.prefix, "-card-pictogram")
  }, /*#__PURE__*/React.createElement(Pictogram, null)), /*#__PURE__*/React.createElement("div", {
    className: "".concat(pkg.prefix, "-card-content-container")
  }, /*#__PURE__*/React.createElement("div", getHeaderBodyProps(), /*#__PURE__*/React.createElement("div", {
    className: headerClasses
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(pkg.prefix, "-card-header-container")
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(pkg.prefix, "-card-title-container")
  }, label && /*#__PURE__*/React.createElement("p", {
    className: "".concat(pkg.prefix, "-card-label")
  }, label), title && /*#__PURE__*/React.createElement("p", {
    className: "".concat(pkg.prefix, "-card-title")
  }, title), caption && /*#__PURE__*/React.createElement("p", {
    className: "".concat(pkg.prefix, "-card-caption")
  }, caption)), hasTopActions && /*#__PURE__*/React.createElement("div", {
    className: "".concat(pkg.prefix, "-card-actions ").concat(pkg.prefix, "-card-actions--top")
  }, getActions()))), /*#__PURE__*/React.createElement("div", getBodyProps(), children)), hasBottomBar && /*#__PURE__*/React.createElement("div", {
    className: footerClasses
  }, secondaryButtonText && /*#__PURE__*/React.createElement(Button, {
    kind: secondaryButtonKind,
    onClick: onSecondaryButtonClick,
    size: "field"
  }, secondaryButtonText), primaryButtonText && /*#__PURE__*/React.createElement(Button, {
    kind: productive ? 'ghost' : primaryButtonKind,
    onClick: onPrimaryButtonClick,
    size: "field"
  }, primaryButtonText), hasBottomActions && /*#__PURE__*/React.createElement("div", {
    className: "".concat(pkg.prefix, "-card-actions")
  }, getActions()))));
  return CardContent;
}; // Return a placeholder if not released and not enabled by feature flag

Card = pkg.checkComponentEnabled(Card, componentName);
Card.propTypes = {
  actionIcons: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    icon: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    onKeyDown: PropTypes.func,
    onClick: PropTypes.func,
    iconDescription: PropTypes.string
  })),
  actionIconsPosition: PropTypes.oneOf(['top', 'bottom']),
  caption: PropTypes.string,
  children: PropTypes.node,
  className: PropTypes.string,
  clickZone: PropTypes.oneOf(['one', 'two', 'three']),
  label: PropTypes.string,
  media: PropTypes.node,
  mediaPosition: PropTypes.oneOf(['top', 'left']),
  onClick: PropTypes.func,
  onKeyDown: PropTypes.func,
  onPrimaryButtonClick: PropTypes.func,
  onSecondaryButtonClick: PropTypes.func,
  overflowActions: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    itemText: PropTypes.string,
    onClick: PropTypes.func,
    onKeyDown: PropTypes.func
  })),
  pictogram: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
  primaryButtonKind: PropTypes.oneOf(['primary', 'ghost']),
  primaryButtonText: PropTypes.string,
  productive: PropTypes.bool,
  secondaryButtonKind: PropTypes.oneOf(['secondary', 'ghost']),
  secondaryButtonText: PropTypes.string,
  title: PropTypes.string,
  titleSize: PropTypes.oneOf(['default', 'large'])
};
Card.defaultProps = {
  actionIcons: [],
  actionIconsPosition: 'bottom',
  clickZone: 'one',
  mediaPosition: 'top',
  overflowActions: [],
  primaryButtonKind: 'primary',
  productive: false,
  secondaryButtonKind: 'secondary',
  titleSize: 'default'
};
Card.displayName = componentName;
