// @carbon/ibm-cloud-cognitive/es/components/ButtonSetWithOverflow/ButtonSetWithOverflow.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
var _excluded = ["buttons", "size"],
    _excluded2 = ["label"],
    _excluded3 = ["buttons"],
    _excluded4 = ["label", "kind"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

//
// Copyright IBM Corp. 2020, 2021
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//
import React, { useEffect, useState, useRef } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import ReactResizeDetector from 'react-resize-detector';
import { ButtonSet, Button } from 'carbon-components-react';
import { ButtonMenu, ButtonMenuItem } from '../ButtonMenu';
import { pkg, carbon } from '../../settings';
import { deprecateProp, extractShapesArray, prepareProps } from '../../global/js/utils/props-helper';
var blockClass = "".concat(pkg.prefix, "--button-set-with-overflow");
var componentName = 'ButtonSetWithOverflow';
export var ButtonSetWithOverflow = function ButtonSetWithOverflow(_ref) {
  var buttons = _ref.buttons,
      children = _ref.children,
      className = _ref.className,
      onWidthChange = _ref.onWidthChange,
      pageActionsLabel = _ref.pageActionsLabel,
      rightAlign = _ref.rightAlign,
      size = _ref.size;

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      showAsOverflow = _useState2[0],
      setShowAsOverflow = _useState2[1];

  var spaceAvailableRef = useRef(null);
  var sizingContainerRefSet = useRef(null);
  var sizingContainerRefCombo = useRef(null);

  var _useState3 = useState([]),
      _useState4 = _slicedToArray(_useState3, 2),
      itemArray = _useState4[0],
      setItemArray = _useState4[1]; // determine display count based on space available and width of pageActions


  var checkFullyVisibleItems = function checkFullyVisibleItems() {
    var _spaceAvailableRef$cu, _sizingContainerRefSe, _sizingContainerRefCo, _sizingContainerRefCo2;

    var spaceAvailable = (_spaceAvailableRef$cu = spaceAvailableRef.current) === null || _spaceAvailableRef$cu === void 0 ? void 0 : _spaceAvailableRef$cu.offsetWidth;
    var newShowAsOverflow = true;
    var sizingSet = (_sizingContainerRefSe = sizingContainerRefSet.current) === null || _sizingContainerRefSe === void 0 ? void 0 : _sizingContainerRefSe.querySelectorAll(".".concat(carbon.prefix, "--btn"));
    var sizingSetTotalSize = 0;

    var _iterator = _createForOfIteratorHelper(sizingSet),
        _step;

    try {
      for (_iterator.s(); !(_step = _iterator.n()).done;) {
        var item = _step.value;
        sizingSetTotalSize += item.offsetWidth;
      }
    } catch (err) {
      _iterator.e(err);
    } finally {
      _iterator.f();
    }

    var sizingComboSize = (_sizingContainerRefCo = sizingContainerRefCombo.current) !== null && _sizingContainerRefCo !== void 0 && _sizingContainerRefCo.offsetWidth ? (_sizingContainerRefCo2 = sizingContainerRefCombo.current) === null || _sizingContainerRefCo2 === void 0 ? void 0 : _sizingContainerRefCo2.offsetWidth : 0;
    onWidthChange && onWidthChange({
      maxWidth: sizingSetTotalSize,
      minWidth: sizingComboSize
    });

    if (sizingSetTotalSize <= spaceAvailable) {
      newShowAsOverflow = false;
    }

    setShowAsOverflow(newShowAsOverflow);
  };

  useEffect(function () {
    checkFullyVisibleItems(); // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [itemArray]);

  var handleResize = function handleResize() {
    // width is the space available for all action bar items horizontally
    // the action bar items are squares so the height should be one item wide

    /* istanbul ignore next */
    // not sure how to test resize
    checkFullyVisibleItems();
  };

  var handleButtonResize = function handleButtonResize() {
    /* istanbul ignore next */
    // not sure how to test resize
    checkFullyVisibleItems();
  };

  useEffect(function () {
    if (buttons) {
      setItemArray(buttons);
    } else {
      var _extractShapesArray;

      setItemArray((_extractShapesArray = extractShapesArray(children)) === null || _extractShapesArray === void 0 ? void 0 : _extractShapesArray.map(function (shape) {
        return _objectSpread({
          label: shape.children
        }, shape);
      }));
    }
  }, [buttons, children]);
  var AButtonSet = /*#__PURE__*/React.forwardRef(function (_ref2, ref) {
    var buttons = _ref2.buttons,
        size = _ref2.size,
        rest = _objectWithoutProperties(_ref2, _excluded);

    return /*#__PURE__*/React.createElement(ButtonSet, _extends({}, rest, {
      ref: ref
    }), buttons.map(function (_ref3, index) {
      var label = _ref3.label,
          other = _objectWithoutProperties(_ref3, _excluded2);

      return /*#__PURE__*/React.createElement(Button, _extends({
        key: index
      }, other, {
        size: size,
        type: "button"
      }), label);
    }));
  });
  var AButtonMenu = /*#__PURE__*/React.forwardRef(function (_ref4, ref) {
    var buttons = _ref4.buttons,
        rest = _objectWithoutProperties(_ref4, _excluded3);

    return /*#__PURE__*/React.createElement(ButtonMenu, _extends({}, rest, {
      ref: ref,
      label: pageActionsLabel
    }), buttons.map(function (_ref5, index) {
      var label = _ref5.label,
          kind = _ref5.kind,
          other = _objectWithoutProperties(_ref5, _excluded4);

      return /*#__PURE__*/React.createElement(ButtonMenuItem, _extends({
        key: index,
        isDelete: kind === 'danger',
        itemText: label
      }, prepareProps(other, ['iconDescription', 'renderIcon'])));
    }).reverse());
  });
  return /*#__PURE__*/React.createElement(ReactResizeDetector, {
    handleWidth: true,
    onResize: handleResize
  }, /*#__PURE__*/React.createElement("div", {
    className: cx([blockClass, className, _defineProperty({}, "".concat(blockClass, "--right"), rightAlign)]),
    ref: spaceAvailableRef
  }, /*#__PURE__*/React.createElement(ReactResizeDetector, {
    onResize: handleButtonResize
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(blockClass, "__button-container ").concat(blockClass, "__button-container--hidden")
  }, /*#__PURE__*/React.createElement(AButtonSet, {
    "aria-hidden": true,
    ref: sizingContainerRefSet,
    size: size,
    buttons: itemArray
  }))), /*#__PURE__*/React.createElement(ReactResizeDetector, {
    onResize: handleButtonResize
  }, /*#__PURE__*/React.createElement("div", {
    className: "".concat(blockClass, "__button-container ").concat(blockClass, "__button-container--hidden"),
    "aria-hidden": true
  }, /*#__PURE__*/React.createElement(AButtonMenu, {
    ref: sizingContainerRefCombo,
    buttons: itemArray,
    size: size
  }))), showAsOverflow ? /*#__PURE__*/React.createElement(AButtonMenu, {
    buttons: itemArray,
    size: size
  }) : /*#__PURE__*/React.createElement(AButtonSet, {
    className: "".concat(blockClass, "__button-container"),
    size: size,
    buttons: itemArray
  })));
};
ButtonSetWithOverflow.propTypes = {
  /**
  /**
   * Specifies the buttons for the ButtonSetWithOverflow. Each item is specified as an object
   * with the properties of a Carbon Button plus a label.
   *
   * Carbon Button API https://react.carbondesignsystem.com/?path=/docs/components-button--default#component-api
   */
  buttons: PropTypes.arrayOf(PropTypes.shape(_objectSpread(_objectSpread({}, Button.propTypes), {}, {
    kind: Button.propTypes.kind,
    label: PropTypes.node,
    onClick: Button.propTypes.onClick
  }))),

  /**
   * children of the button set
   */
  children: deprecateProp(PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.element), PropTypes.element]), 'See documentation on the `buttons` property.'),
  // expects action bar item as array or in fragment,

  /**
   * className
   */
  className: PropTypes.string,

  /**
   * onResize reports maxSize on resize
   */
  onWidthChange: PropTypes.func,

  /**
   * pageActionsLabel - used when button set is shown as combo button
   */
  pageActionsLabel: PropTypes.node,

  /**
   * align buttons to right of available space
   */
  rightAlign: PropTypes.bool,

  /**
   * Specify the size of the button, from a list of available sizes.
   * For `default` buttons, this prop can remain unspecified.
   */
  size: Button.propTypes.size
};
ButtonSetWithOverflow.defaultProps = {
  pageActionsLabel: 'Page actions',
  size: 'field'
};
ButtonSetWithOverflow.displayName = componentName;
