// @carbon/ibm-cloud-cognitive/lib/components/BreadcrumbWithOverflow/BreadcrumbWithOverflow.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BreadcrumbWithOverflow = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _carbonComponentsReact = require("carbon-components-react");

var _settings = require("../../settings");

var _reactResizeDetector = _interopRequireDefault(require("react-resize-detector"));

var _iconsReact = require("@carbon/icons-react");

var _uuidv = _interopRequireDefault(require("../../global/js/utils/uuidv4"));

var _unwrapIfFragment = _interopRequireDefault(require("../../global/js/utils/unwrap-if-fragment"));

var _excluded = ["children", "className", "maxVisible", "noTrailingSlash", "overflowAriaLabel"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

// The block part of our conventional BEM class names (blockClass__E--M).
var blockClass = "".concat(_settings.pkg.prefix, "--breadcrumb-with-overflow");
var componentName = 'BreadcrumbWithOverflow'; // NOTE: the component SCSS is not imported here: it is rolled up separately.

/**
 * The BreadcrumbWithOverflow is used internally by the PageHeader to wrap BreadcrumbItems.
 */

var BreadcrumbWithOverflow = function BreadcrumbWithOverflow(_ref) {
  var _backItem, _backItem$props;

  var children = _ref.children,
      className = _ref.className,
      maxVisible = _ref.maxVisible,
      noTrailingSlash = _ref.noTrailingSlash,
      overflowAriaLabel = _ref.overflowAriaLabel,
      other = (0, _objectWithoutProperties2.default)(_ref, _excluded);

  var _useState = (0, _react.useState)(3),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      displayCount = _useState2[0],
      setDisplayCount = _useState2[1];

  var _useState3 = (0, _react.useState)([]),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      displayedBreadcrumbItems = _useState4[0],
      setDisplayedBreadcrumbItems = _useState4[1];

  var breadcrumbItemWithOverflow = (0, _react.useRef)(null);
  var sizingContainerRef = (0, _react.useRef)(null);
  var internalId = (0, _react.useRef)((0, _uuidv.default)());

  var _useState5 = (0, _react.useState)([]),
      _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
      childArray = _useState6[0],
      setChildArray = _useState6[1];

  var getHref = function getHref(item) {
    var _item$props, _item$props2, _item$props2$children, _item$props2$children2;

    // This function should extract href from item
    // It expects that the href is attached either to the item or direct child
    // It prefers item.props.href
    return item !== null && item !== void 0 && (_item$props = item.props) !== null && _item$props !== void 0 && _item$props.href ? item.props.href : item === null || item === void 0 ? void 0 : (_item$props2 = item.props) === null || _item$props2 === void 0 ? void 0 : (_item$props2$children = _item$props2.children) === null || _item$props2$children === void 0 ? void 0 : (_item$props2$children2 = _item$props2$children.props) === null || _item$props2$children2 === void 0 ? void 0 : _item$props2$children2.href;
  };

  var getTitle = function getTitle(item) {
    // This function should extract text based title from the item.
    // It prefers in this order
    // - item.props.data-title
    // - item.props.title
    // - item.props.children if string
    // - item.props.children.props.children if string. This case is likely if an <a /> is used inside a BreadcrumbItem
    var useAsTitle = null;

    if (item !== null && item !== void 0 && item.props) {
      var _item$props3, _item$props3$children, _item$props3$children2;

      if (item.props['data-title']) {
        useAsTitle = item.props['data-title'];
      } else if (item.props.title) {
        useAsTitle = item.props.title;
      } else if (typeof item.props.children === 'string') {
        useAsTitle = item.props.children;
      } else if (typeof ((_item$props3 = item.props) === null || _item$props3 === void 0 ? void 0 : (_item$props3$children = _item$props3.children) === null || _item$props3$children === void 0 ? void 0 : (_item$props3$children2 = _item$props3$children.props) === null || _item$props3$children2 === void 0 ? void 0 : _item$props3$children2.children) === 'string') {
        useAsTitle = item.props.children.props.children;
      }
    }

    return useAsTitle;
  }; // eslint-disable-next-line react/prop-types


  var BreadcrumbOverflowMenu = function BreadcrumbOverflowMenu(_ref2) {
    var overflowItems = _ref2.overflowItems;
    return /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.BreadcrumbItem, {
      key: "breadcrumb-overflow-".concat(internalId.current)
    }, /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.OverflowMenu, {
      ariaLabel: null,
      menuOffset: {
        top: 10,
        left: 59
      } // TODO: REMOVE borrowed from https://github.com/carbon-design-system/carbon/pull/7085
      ,
      renderIcon: _iconsReact.OverflowMenuHorizontal32,
      className: "".concat(blockClass, "__overflow-menu"),
      menuOptionsClass: "".concat(_settings.carbon.prefix, "--breadcrumb-menu-options") // TODO: REMOVE borrowed from https://github.com/carbon-design-system/carbon/pull/7085

    }, // eslint-disable-next-line react/prop-types
    overflowItems.map(function (item, index) {
      return /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.OverflowMenuItem, {
        key: "breadcrumb-overflow-menu-item-".concat(internalId.current, "-").concat(index),
        href: item.props.href,
        onClick: item.props.onClick,
        itemText: item.props.children
      });
    })));
  }; // creates child array from children which may be a fragment


  (0, _react.useEffect)(function () {
    setChildArray((0, _unwrapIfFragment.default)(children));
  }, [children]);
  (0, _react.useEffect)(function () {
    // updates displayedBreadcrumbItems and overflowBreadcrumbItems based on displayCount and childArray
    if (childArray.length === 0) {
      setDisplayedBreadcrumbItems([]);
      return;
    } // clones of children needed as the children are used in the sizing render


    var cloneChildren = function cloneChildren(items) {
      return items.map(function (item, index) {
        // likely truncated add title
        var title = index + 1 === childArray.length && displayCount === 1 ? getTitle(item) : null;
        var className = index > 0 || displayCount > 1 ? (0, _classnames.default)([childArray[index].props.className, "".concat(blockClass, "__displayed-breadcrumb")]) : childArray[index].props.className;
        return /*#__PURE__*/_react.default.cloneElement(item, {
          key: index,
          title: title,
          className: className
        });
      });
    };

    var newDisplayedBreadcrumbItems = cloneChildren(childArray); // The breadcrumb has the form [first item] [overflow] [items 2...(n-1)] [last item].
    // The overflow is only shown if there isn't space to display all the items, and in that case:
    //  * the last item is always displayed (even if there isn't really space for it -- it can contract to an ellipsis);
    //  * the first item is the next to be displayed, if there's space once the last item and overflow are shown;
    //  * any remaining space after the first item, last item and overflow are shown is used to show items (n-1), (n-2), (n-3), ..., until the space is used up ;
    // Note that displayCount (min 1) has been computed based on the available space and the above sequence.

    var overflowPosition = displayCount > 1 ? 1 : 0;
    var newOverflowBreadcrumbItems = newDisplayedBreadcrumbItems.splice(overflowPosition, childArray.length - displayCount); // if needed add overflow menu

    if (newOverflowBreadcrumbItems.length) {
      newDisplayedBreadcrumbItems.splice(overflowPosition, 0, /*#__PURE__*/_react.default.createElement(BreadcrumbOverflowMenu, {
        overflowItems: newOverflowBreadcrumbItems,
        key: "$displayed-breadcrumb-".concat(internalId, "-overflow")
      }));
    }

    setDisplayedBreadcrumbItems(newDisplayedBreadcrumbItems);
  }, [childArray, displayCount]);

  var checkFullyVisibleBreadcrumbItems = function checkFullyVisibleBreadcrumbItems() {
    var displayItemIndex = function displayItemIndex(itemCount, index) {
      // In this data set the overflow measuring item is [0]
      // so the first displayItem in the list is [1]
      // we never return 0;
      if (index === 0) {
        return itemCount - 1; // the last item in the list
      } else if (index === 1) {
        return 1; // the first item in the list
      } else {
        return itemCount - index; // count down from itemCount - 2 to 1
      }
    };

    if (maxVisible <= 1) {
      setDisplayCount(1);
    } else {
      // how many will fit?
      var willFit = 0;
      var spaceAvailable = breadcrumbItemWithOverflow.current.offsetWidth; // not sure how to test resize

      /* istanbul ignore next if */

      if (sizingContainerRef.current) {
        var sizingBreadcrumbItems = sizingContainerRef.current.querySelectorAll(".".concat(_settings.carbon.prefix, "--breadcrumb-item"));
        var breadcrumbWidthsIncludingMargin = [];

        var _iterator = _createForOfIteratorHelper(sizingBreadcrumbItems),
            _step;

        try {
          for (_iterator.s(); !(_step = _iterator.n()).done;) {
            var item = _step.value;
            var computedStyle = window ? window.getComputedStyle(sizingBreadcrumbItems[0]) : null;
            var marginWidths = computedStyle ? parseFloat(computedStyle.marginLeft || 0, 10) + parseFloat(computedStyle.marginRight || 0, 10) : 0;
            breadcrumbWidthsIncludingMargin.push(item.offsetWidth + marginWidths);
          }
        } catch (err) {
          _iterator.e(err);
        } finally {
          _iterator.f();
        }

        var overflowWidth = breadcrumbWidthsIncludingMargin[0];

        for (var i = 0; i < breadcrumbWidthsIncludingMargin.length - 1; i++) {
          // count used one less than length to account for the included overflow item
          var index = displayItemIndex(breadcrumbWidthsIncludingMargin.length, i);

          if (spaceAvailable >= breadcrumbWidthsIncludingMargin[index]) {
            spaceAvailable -= breadcrumbWidthsIncludingMargin[index];
            willFit += 1;
          } else {
            break;
          }
        } // if not enough space for all breadcrumb items


        if (willFit < breadcrumbWidthsIncludingMargin.length - 1) {
          // -1 for overflow item
          while (willFit > 0 && spaceAvailable < overflowWidth) {
            willFit -= 1; // Highly unlikely any useful breadcrumb-item is smaller than the overflow menu, but we loop anyway just in case
            // item removed is based on last item added which is the current value of willFit

            var itemToRemove = displayItemIndex(breadcrumbWidthsIncludingMargin.length, willFit);
            spaceAvailable += breadcrumbWidthsIncludingMargin[itemToRemove];
          }
        }
      }

      if (willFit <= 1) {
        setDisplayCount(1);
      } else {
        setDisplayCount(maxVisible ? Math.min(willFit, maxVisible) : willFit);
      }
    }
  };

  (0, _react.useEffect)(function () {
    checkFullyVisibleBreadcrumbItems(); // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [maxVisible]);

  var handleResize = function handleResize() {
    /* istanbul ignore next */
    // not sure how to test resize
    checkFullyVisibleBreadcrumbItems();
  };

  var handleBreadcrumbItemsResize = function handleBreadcrumbItemsResize() {
    /* istanbul ignore next */
    // not sure how to test resize
    checkFullyVisibleBreadcrumbItems();
  };

  var backItem = childArray[childArray.length - 1];

  if ((_backItem = backItem) !== null && _backItem !== void 0 && (_backItem$props = _backItem.props) !== null && _backItem$props !== void 0 && _backItem$props.isCurrentPage) {
    backItem = childArray[childArray.length - 2];
  }

  var buttonHrefValue = getHref(backItem);
  var buttonTooltipValue = getTitle(backItem);
  return /*#__PURE__*/_react.default.createElement(_reactResizeDetector.default, {
    onResize: handleResize
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)(blockClass, className, (0, _defineProperty2.default)({}, "".concat(blockClass, "__with-items"), displayedBreadcrumbItems.length > 1)),
    ref: breadcrumbItemWithOverflow
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)(["".concat(blockClass, "__space")])
  }, /*#__PURE__*/_react.default.createElement(_reactResizeDetector.default, {
    onResize: handleBreadcrumbItemsResize
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(blockClass, "__breadcrumb-container ").concat(blockClass, "__breadcrumb-container--hidden"),
    "aria-hidden": true,
    ref: sizingContainerRef
  }, /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.Breadcrumb, null, /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.BreadcrumbItem, {
    key: "".concat(blockClass, "-hidden-overflow-").concat(internalId)
  }, /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.OverflowMenu, {
    ariaLabel: overflowAriaLabel,
    renderIcon: _iconsReact.OverflowMenuHorizontal32
  })), children))), buttonHrefValue && buttonTooltipValue && /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.Button, {
    className: "".concat(blockClass, "--breadcrumb-back-button"),
    hasIconOnly: true,
    iconDescription: buttonTooltipValue,
    kind: "ghost",
    href: buttonHrefValue || '#',
    renderIcon: _iconsReact.ArrowLeft16,
    size: "field",
    tooltipPosition: "right",
    type: "button"
  }), /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.Breadcrumb, (0, _extends2.default)({
    className: (0, _classnames.default)("".concat(blockClass, "__breadcrumb-container"), (0, _defineProperty2.default)({}, "".concat(blockClass, "__breadcrumb-container-with-items"), displayedBreadcrumbItems.length > 1)),
    noTrailingSlash: noTrailingSlash
  }, other), displayedBreadcrumbItems))));
}; // Return a placeholder if not released and not enabled by feature flag


exports.BreadcrumbWithOverflow = BreadcrumbWithOverflow;
exports.BreadcrumbWithOverflow = BreadcrumbWithOverflow = _settings.pkg.checkComponentEnabled(BreadcrumbWithOverflow, componentName);
BreadcrumbWithOverflow.propTypes = {
  /**
   * children of the breadcrumb-item set (these are expected to be breadcrumb-items)
   */
  children: _propTypes.default.arrayOf(_propTypes.default.element),

  /**
   * className
   */
  className: _propTypes.default.string,

  /**
   * maxVisble: maximum visible breadcrumb-items before overflow is used (values less than 1 are treated as 1)
   */
  maxVisible: _propTypes.default.number,

  /**
   * noTrailing slash - same as for Carbon
   */
  noTrailingSlash: _propTypes.default.bool,

  /**
   * overflowAriaLabel label for open close button overflow used for action bar items that do nto fit.
   */
  overflowAriaLabel: _propTypes.default.string
};
BreadcrumbWithOverflow.defaultProps = {
  overflowAriaLabel: 'Open and close additional breadcrumb item list.',
  noTrailingSlash: false
};
BreadcrumbWithOverflow.displayName = componentName;
