// @carbon/ibm-cloud-cognitive/lib/components/Card/Card.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Card = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _carbonComponentsReact = require("carbon-components-react");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _settings = require("../../settings");

var _excluded = ["id"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var componentName = 'Card';

var Card = function Card(_ref) {
  var _cx;

  var actionIcons = _ref.actionIcons,
      actionIconsPosition = _ref.actionIconsPosition,
      caption = _ref.caption,
      children = _ref.children,
      className = _ref.className,
      clickZone = _ref.clickZone,
      label = _ref.label,
      media = _ref.media,
      mediaPosition = _ref.mediaPosition,
      onClick = _ref.onClick,
      onKeyDown = _ref.onKeyDown,
      onPrimaryButtonClick = _ref.onPrimaryButtonClick,
      overflowActions = _ref.overflowActions,
      onSecondaryButtonClick = _ref.onSecondaryButtonClick,
      Pictogram = _ref.pictogram,
      primaryButtonKind = _ref.primaryButtonKind,
      primaryButtonText = _ref.primaryButtonText,
      productive = _ref.productive,
      secondaryButtonKind = _ref.secondaryButtonKind,
      secondaryButtonText = _ref.secondaryButtonText,
      title = _ref.title,
      titleSize = _ref.titleSize;
  var hasActions = actionIcons.length > 0 || overflowActions.length > 0;
  var hasTopActions = hasActions && actionIconsPosition === 'top';
  var hasBottomActions = hasActions && actionIconsPosition === 'bottom';
  var hasBottomButton = !!secondaryButtonText || !!primaryButtonText;
  var hasBottomBar = hasBottomActions || hasBottomButton;
  var headerClasses = (0, _classnames.default)("".concat(_settings.pkg.prefix, "-card-header"), (_cx = {}, (0, _defineProperty2.default)(_cx, "".concat(_settings.pkg.prefix, "-card-header--label-only"), label && !title && !caption), (0, _defineProperty2.default)(_cx, "".concat(_settings.pkg.prefix, "-card-header--has-label"), label && productive), (0, _defineProperty2.default)(_cx, "".concat(_settings.pkg.prefix, "-card-title--lg"), titleSize === 'large'), _cx));
  var footerClasses = (0, _classnames.default)("".concat(_settings.pkg.prefix, "-card-footer"), (0, _defineProperty2.default)({}, "".concat(_settings.pkg.prefix, "-card-footer--no-button"), !hasBottomButton));

  var getActions = function getActions() {
    if (overflowActions.length !== 0) {
      var pos = actionIconsPosition === 'top' ? 'bottom' : 'top';
      var size = actionIconsPosition === 'top' ? 'sm' : 'xl';
      return /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.OverflowMenu, {
        size: size,
        direction: pos,
        flipped: true
      }, overflowActions.map(function (_ref2) {
        var id = _ref2.id,
            rest = (0, _objectWithoutProperties2.default)(_ref2, _excluded);
        return /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.OverflowMenuItem, (0, _extends2.default)({
          key: id
        }, rest));
      }));
    }

    var icons = actionIcons.map(function (_ref3) {
      var id = _ref3.id,
          Icon = _ref3.icon,
          onClick = _ref3.onClick,
          iconDescription = _ref3.iconDescription,
          onKeyDown = _ref3.onKeyDown;

      if (productive) {
        return /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.Button, {
          key: id,
          renderIcon: Icon,
          hasIconOnly: true,
          onClick: onClick,
          size: actionIconsPosition === 'top' ? 'sm' : 'field',
          iconDescription: iconDescription,
          kind: "ghost"
        });
      }

      return /*#__PURE__*/_react.default.createElement("div", {
        key: id,
        className: "".concat(_settings.pkg.prefix, "-card-icon"),
        onClick: onClick,
        onKeyDown: onKeyDown,
        role: "button",
        tabIndex: "0"
      }, /*#__PURE__*/_react.default.createElement(Icon, {
        "aria-label": iconDescription
      }));
    });
    return icons;
  };

  var getClickableProps = function getClickableProps() {
    return {
      onClick: onClick,
      onKeyDown: onKeyDown,
      role: 'button',
      tabIndex: '0'
    };
  };

  var getCardProps = function getCardProps() {
    var _cx3;

    var hasClickEvent = !!onClick || !!onKeyDown;
    var clickable = hasClickEvent && !productive || hasClickEvent && productive && clickZone === 'one';

    var cardProps = _objectSpread({
      className: (0, _classnames.default)("".concat(_settings.pkg.prefix, "-card"), (_cx3 = {}, (0, _defineProperty2.default)(_cx3, "".concat(_settings.pkg.prefix, "-card--productive"), productive), (0, _defineProperty2.default)(_cx3, "".concat(_settings.pkg.prefix, "-card--clickable"), clickable), (0, _defineProperty2.default)(_cx3, "".concat(_settings.pkg.prefix, "-card--media-left"), mediaPosition === 'left'), (0, _defineProperty2.default)(_cx3, "className", className), _cx3))
    }, clickable && getClickableProps());

    return cardProps;
  };

  var getHeaderBodyProps = function getHeaderBodyProps() {
    var hasClickEvent = !!onClick || !!onKeyDown;
    var clickable = hasClickEvent && clickZone === 'two';

    var headerBodyProps = _objectSpread({
      className: (0, _classnames.default)("".concat(_settings.pkg.prefix, "-card-header-body-container"), (0, _defineProperty2.default)({}, "".concat(_settings.pkg.prefix, "-card--clickable"), clickable))
    }, clickable && getClickableProps());

    return headerBodyProps;
  };

  var getBodyProps = function getBodyProps() {
    var hasClickEvent = !!onClick || !!onKeyDown;
    var clickable = hasClickEvent && clickZone === 'three';

    var bodyProps = _objectSpread({
      className: (0, _classnames.default)("".concat(_settings.pkg.prefix, "-card-body"), (0, _defineProperty2.default)({}, "".concat(_settings.pkg.prefix, "-card--clickable"), clickable))
    }, clickable && getClickableProps());

    return bodyProps;
  };

  var CardContent = /*#__PURE__*/_react.default.createElement("div", getCardProps(), media && /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(_settings.pkg.prefix, "-card-media")
  }, media), Pictogram && /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(_settings.pkg.prefix, "-card-pictogram")
  }, /*#__PURE__*/_react.default.createElement(Pictogram, null)), /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(_settings.pkg.prefix, "-card-content-container")
  }, /*#__PURE__*/_react.default.createElement("div", getHeaderBodyProps(), /*#__PURE__*/_react.default.createElement("div", {
    className: headerClasses
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(_settings.pkg.prefix, "-card-header-container")
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(_settings.pkg.prefix, "-card-title-container")
  }, label && /*#__PURE__*/_react.default.createElement("p", {
    className: "".concat(_settings.pkg.prefix, "-card-label")
  }, label), title && /*#__PURE__*/_react.default.createElement("p", {
    className: "".concat(_settings.pkg.prefix, "-card-title")
  }, title), caption && /*#__PURE__*/_react.default.createElement("p", {
    className: "".concat(_settings.pkg.prefix, "-card-caption")
  }, caption)), hasTopActions && /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(_settings.pkg.prefix, "-card-actions ").concat(_settings.pkg.prefix, "-card-actions--top")
  }, getActions()))), /*#__PURE__*/_react.default.createElement("div", getBodyProps(), children)), hasBottomBar && /*#__PURE__*/_react.default.createElement("div", {
    className: footerClasses
  }, secondaryButtonText && /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.Button, {
    kind: secondaryButtonKind,
    onClick: onSecondaryButtonClick,
    size: "field"
  }, secondaryButtonText), primaryButtonText && /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.Button, {
    kind: productive ? 'ghost' : primaryButtonKind,
    onClick: onPrimaryButtonClick,
    size: "field"
  }, primaryButtonText), hasBottomActions && /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(_settings.pkg.prefix, "-card-actions")
  }, getActions()))));

  return CardContent;
}; // Return a placeholder if not released and not enabled by feature flag


exports.Card = Card;
exports.Card = Card = _settings.pkg.checkComponentEnabled(Card, componentName);
Card.propTypes = {
  actionIcons: _propTypes.default.arrayOf(_propTypes.default.shape({
    id: _propTypes.default.string,
    icon: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object]),
    onKeyDown: _propTypes.default.func,
    onClick: _propTypes.default.func,
    iconDescription: _propTypes.default.string
  })),
  actionIconsPosition: _propTypes.default.oneOf(['top', 'bottom']),
  caption: _propTypes.default.string,
  children: _propTypes.default.node,
  className: _propTypes.default.string,
  clickZone: _propTypes.default.oneOf(['one', 'two', 'three']),
  label: _propTypes.default.string,
  media: _propTypes.default.node,
  mediaPosition: _propTypes.default.oneOf(['top', 'left']),
  onClick: _propTypes.default.func,
  onKeyDown: _propTypes.default.func,
  onPrimaryButtonClick: _propTypes.default.func,
  onSecondaryButtonClick: _propTypes.default.func,
  overflowActions: _propTypes.default.arrayOf(_propTypes.default.shape({
    id: _propTypes.default.string,
    itemText: _propTypes.default.string,
    onClick: _propTypes.default.func,
    onKeyDown: _propTypes.default.func
  })),
  pictogram: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object]),
  primaryButtonKind: _propTypes.default.oneOf(['primary', 'ghost']),
  primaryButtonText: _propTypes.default.string,
  productive: _propTypes.default.bool,
  secondaryButtonKind: _propTypes.default.oneOf(['secondary', 'ghost']),
  secondaryButtonText: _propTypes.default.string,
  title: _propTypes.default.string,
  titleSize: _propTypes.default.oneOf(['default', 'large'])
};
Card.defaultProps = {
  actionIcons: [],
  actionIconsPosition: 'bottom',
  clickZone: 'one',
  mediaPosition: 'top',
  overflowActions: [],
  primaryButtonKind: 'primary',
  productive: false,
  secondaryButtonKind: 'secondary',
  titleSize: 'default'
};
Card.displayName = componentName;
