// @carbon/ibm-cloud-cognitive/lib/components/ProductiveCard/ProductiveCard.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ProductiveCard = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Card = require("../Card");

var _settings = require("../../settings");

var _propsHelper = require("../../global/js/utils/props-helper");

//
// Copyright IBM Corp. 2020, 2021
//
// This source code is licensed under the Apache-2.0 license found in the
// LICENSE file in the root directory of this source tree.
//
var componentName = 'ProductiveCard';

var ProductiveCard = function ProductiveCard(props) {
  var validProps = (0, _propsHelper.prepareProps)(props, ['media', 'mediaPosition', 'onSecondaryButtonClick', 'pictogram', 'primaryButtonClick', 'productive', 'secondaryButtonKind', 'secondaryButtonText']);
  return /*#__PURE__*/_react.default.createElement(_Card.Card, (0, _extends2.default)({}, validProps, {
    productive: true
  }));
}; // Return a placeholder if not released and not enabled by feature flag


exports.ProductiveCard = ProductiveCard;
exports.ProductiveCard = ProductiveCard = _settings.pkg.checkComponentEnabled(ProductiveCard, componentName);
ProductiveCard.propTypes = {
  /**
   * Icons that are displayed on card. Refer to design documentation for implementation guidelines
   */
  actionIcons: _propTypes.default.arrayOf(_propTypes.default.shape({
    id: _propTypes.default.string,
    icon: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object]),
    onKeyDown: _propTypes.default.func,
    onClick: _propTypes.default.func,
    iconDescription: _propTypes.default.string
  })),

  /**
   * Determines if the action icons are on the top or bottom of the card
   */
  actionIconsPosition: _propTypes.default.oneOf(['top', 'bottom']),

  /**
   * Optional header caption
   */
  caption: _propTypes.default.string,

  /**
   * Content that shows in the body of the card
   */
  children: _propTypes.default.node,

  /**
   * Optional user provided class
   */
  className: _propTypes.default.string,

  /**
   * Designates which zones of the card are clickable. Refer to design documentation for implementation guidelines
   */
  clickZone: _propTypes.default.oneOf(['one', 'two', 'three']),

  /**
   * Optional label for the top of the card
   */
  label: _propTypes.default.string,

  /**
   * Provides the callback for a clickable card
   */
  onClick: _propTypes.default.func,

  /**
   * Function that's called from the primary button or action icon
   */
  onPrimaryButtonClick: _propTypes.default.func,

  /**
   * Use an overflow menu instead of action icons. Refer to design documentation for implementation guidelines
   */
  overflowActions: _propTypes.default.arrayOf(_propTypes.default.shape({
    id: _propTypes.default.string,
    itemText: _propTypes.default.string,
    onClick: _propTypes.default.func,
    onKeyDown: _propTypes.default.func
  })),

  /**
   * The text that's displayed in the primary button
   */
  primaryButtonText: _propTypes.default.string,

  /**
   * Title that's displayed at the top of the card
   */
  title: _propTypes.default.string,

  /**
   * Determines title size
   */
  titleSize: _propTypes.default.oneOf(['default', 'large'])
};
ProductiveCard.defaultProps = {
  actionIcons: [],
  actionIconsPosition: 'top',
  clickZone: 'one',
  overflowActions: [],
  titleSize: 'default'
};
ProductiveCard.displayName = componentName;
