// @carbon/ibm-cloud-cognitive/lib/components/CreateFullPage/CreateFullPageSection.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CreateFullPageSection = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _settings = require("../../settings");

var _carbonComponentsReact = require("carbon-components-react");

var _constants = require("./constants");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var componentName = 'CreateFullPageSection';
var blockClass = "".concat(_settings.pkg.prefix, "--create-full-page__section");
var CreateFullPageSection = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var children = _ref.children,
      className = _ref.className,
      _ref$hasDivider = _ref.hasDivider,
      hasDivider = _ref$hasDivider === void 0 ? true : _ref$hasDivider,
      _ref$hasForm = _ref.hasForm,
      hasForm = _ref$hasForm === void 0 ? true : _ref$hasForm,
      title = _ref.title,
      subtitle = _ref.subtitle,
      description = _ref.description;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)(blockClass, className),
    ref: ref
  }, title && /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.Row, null, /*#__PURE__*/_react.default.createElement(_carbonComponentsReact.Column, null, /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(blockClass, "-text-content")
  }, /*#__PURE__*/_react.default.createElement("h2", {
    className: "".concat(blockClass, "-title")
  }, title), subtitle && /*#__PURE__*/_react.default.createElement("h3", {
    className: "".concat(blockClass, "-subtitle")
  }, subtitle), description && /*#__PURE__*/_react.default.createElement("p", {
    className: "".concat(blockClass, "-description")
  }, description)))), hasForm && /*#__PURE__*/_react.default.createElement("div", {
    className: "".concat(blockClass, "-form")
  }, children), hasDivider && /*#__PURE__*/_react.default.createElement("span", {
    className: "".concat(blockClass, "-divider")
  }));
}); // Return a placeholder if not released and not enabled by feature flag

exports.CreateFullPageSection = CreateFullPageSection;
exports.CreateFullPageSection = CreateFullPageSection = _settings.pkg.checkComponentEnabled(CreateFullPageSection, componentName);
CreateFullPageSection.propTypes = {
  /**
   * Content that shows in the CreateFullPage step
   */
  children: _propTypes.default.node,

  /**
   * Sets an optional className to be added to the CreateFullPage step
   */
  className: _propTypes.default.string,

  /**
   * Sets an optional description for a create full page step
   */
  description: _propTypes.default.node,

  /**
   * This will conditionally disable the submit button in the multi step CreateFullPage
   */
  disableSubmit: _propTypes.default.bool,

  /**
   * This will conditionally render a 1px gray divider span. This is defaulted to true
   */
  hasDivider: _propTypes.default.bool,

  /**
   * This will conditionally render form content below the step condition. This is defaulted to true
   */
  hasForm: _propTypes.default.bool,

  /**
   * Sets the subtitle text for a create full page step
   */
  subtitle: _propTypes.default.node,

  /**
   * Sets the title text for a CreateFullPage step
   */
  title: _propTypes.default.node
}; // Default values for component props. Default values are not required for
// props that are required, nor for props where the component can apply
// 'undefined' values reasonably. Default values should be provided when the
// component needs to make a choice or assumption when a prop is not supplied.

CreateFullPageSection.defaultProps = {
  type: _constants.CREATE_FULL_PAGE_SECTION
};
